In [1]:
import tensorflow as tf
import pandas as pd
import numpy as np

# Read the dataset

In [45]:
data = pd.read_csv("churn.csv")

In [46]:
data.shape

(10000, 14)

In [47]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

# Drop the columns which are unique for all users

In [48]:
data.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [49]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
RowNumber          10000 non-null int64
CustomerId         10000 non-null int64
Surname            10000 non-null object
CreditScore        10000 non-null int64
Geography          10000 non-null object
Gender             10000 non-null object
Age                10000 non-null int64
Tenure             10000 non-null int64
Balance            10000 non-null float64
NumOfProducts      10000 non-null int64
HasCrCard          10000 non-null int64
IsActiveMember     10000 non-null int64
EstimatedSalary    10000 non-null float64
Exited             10000 non-null int64
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [50]:
data = data.drop(columns=['RowNumber','CustomerId','Surname']) #Dropping cloumns which are not useful

In [51]:
data.shape

(10000, 11)

In [52]:
data.columns

Index(['CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance',
       'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary',
       'Exited'],
      dtype='object')

# Distinguish the feature and target set

In [53]:
Y = data['Exited']
X = data.drop(columns=['Exited'])

In [54]:
X.head()

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  
0          1               1        101348.88  
1          0               1        112542.58  
2          1               0        113931.57  
3          0               0         93826.63  
4          1               1         79084.10

In [55]:
Y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

There are two categorical columns left after dropping unnecessary columns. We need to convert these 2 columns Gender and geography to numeric so that we can use these in our model.
For this purpose we will use Get Dummies Method of dataframe.

In [56]:
geography = pd.get_dummies(X['Geography'], drop_first=True)
geography.head()

Germany  Spain
0        0      0
1        0      1
2        0      0
3        0      0
4        0      1

In [57]:
gender = pd.get_dummies(X['Gender'], drop_first=True)
gender.head()

Male
0     0
1     0
2     0
3     0
4     0

In [58]:
X = X.drop(columns=['Geography', 'Gender'], axis=1)

In [59]:
X.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  
0               1        101348.88  
1               1        112542.58  
2               0        113931.57  
3               0         93826.63  
4               1         79084.10

In [60]:
X = pd.concat([X, geography, gender], axis=1)

In [61]:
X.head()

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Germany  Spain  Male  
0               1        101348.88        0      0     0  
1               1        112542.58        0      1     0  
2               0        113931.57        0      0     0  
3               0         93826.63        0      0     0  
4               1         79084.10        0      1     0

In [62]:
X.shape

(10000, 11)

# Divide the data set into train and test sets

In [63]:
from sklearn.model_selection import train_test_split

In [64]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.30, random_state=42)

In [65]:
X_train.shape

(7000, 11)

In [66]:
y_train.shape

(7000,)

In [67]:
X_test.shape

(3000, 11)

# Normalize the train and test data

In [68]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [69]:
X_train = X_train.astype('float32')
X_train.dtype

dtype('float32')

In [70]:
X_test = X_test.astype('float32')
X_test.dtype

dtype('float32')

In [71]:
y_train = y_train.astype('float32')
y_train.dtype

dtype('float32')

In [72]:
y_test = y_test.astype('float32')
y_test.dtype

dtype('float32')

In [73]:
y_train = np.asarray(y_train)

In [74]:
y_test = np.asarray(y_test)

# Initialize & build the model

In [75]:
import tensorflow as tf

In [82]:
model = tf.keras.models.Sequential()

We have 11 independent variables so input_dim will be 11.
We will use relu as an activation function in hidden layers.
We are taking 6 neurons in the hidden layer.
For the output layer, we are taking sigmoid as an activation because there will be only 2 output.
We are also taking, binary_crossentropy as a loss because output will be binary.

In [83]:
#Input layer
model.add(tf.keras.layers.Dense(activation = 'relu', input_dim = 11, units=6, kernel_initializer='uniform'))

#output layer
model.add(tf.keras.layers.Dense(1, activation='sigmoid', name='Output'))

#Compiling
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [84]:
model.fit(X_train,y_train,
        validation_data=(X_test,y_test),
        epochs=40)

Train on 7000 samples, validate on 3000 samples
Epoch 1/40
7000/7000 [==============================] - 1s 124us/sample - loss: 0.6484 - accuracy: 0.7499 - val_loss: 0.5861 - val_accuracy: 0.8197
Epoch 2/40
7000/7000 [==============================] - 0s 50us/sample - loss: 0.5372 - accuracy: 0.8117 - val_loss: 0.4846 - val_accuracy: 0.8350
Epoch 3/40
7000/7000 [==============================] - 0s 49us/sample - loss: 0.4658 - accuracy: 0.8249 - val_loss: 0.4306 - val_accuracy: 0.8363
Epoch 4/40
7000/7000 [==============================] - 0s 56us/sample - loss: 0.4300 - accuracy: 0.8341 - val_loss: 0.4013 - val_accuracy: 0.8530
Epoch 5/40
7000/7000 [==============================] - 0s 49us/sample - loss: 0.4077 - accuracy: 0.8396 - val_loss: 0.3835 - val_accuracy: 0.8543
Epoch 6/40
7000/7000 [==============================] - 0s 64us/sample - loss: 0.3939 - accuracy: 0.8434 - val_loss: 0.3726 - val_accuracy: 0.8617
Epoch 7/40
7000/7000 [==============================] - 0s 56us/sampl

In [85]:
model.summary()

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_19 (Dense)             (None, 6)                 72        
_________________________________________________________________
Output (Dense)               (None, 1)                 7         
Total params: 79
Trainable params: 79
Non-trainable params: 0
_________________________________________________________________


# Optimize the model

In [92]:
model2 = tf.keras.models.Sequential()

#Input layer
model2.add(tf.keras.layers.Dense(activation = 'relu', input_dim = 11, units=8, kernel_initializer='uniform'))

#output layer
model2.add(tf.keras.layers.Dense(1, activation='sigmoid', name='Output'))

#Compiling
model2.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [93]:
model2.fit(X_train,y_train,validation_data=(X_test,y_test),epochs=100, batch_size=32)

Train on 7000 samples, validate on 3000 samples
Epoch 1/100
7000/7000 [==============================] - 1s 91us/sample - loss: 0.5502 - accuracy: 0.7900 - val_loss: 0.4506 - val_accuracy: 0.8073
Epoch 2/100
7000/7000 [==============================] - 0s 50us/sample - loss: 0.4420 - accuracy: 0.8033 - val_loss: 0.4108 - val_accuracy: 0.8200
Epoch 3/100
7000/7000 [==============================] - 0s 52us/sample - loss: 0.4218 - accuracy: 0.8183 - val_loss: 0.3975 - val_accuracy: 0.8320
Epoch 4/100
7000/7000 [==============================] - 0s 48us/sample - loss: 0.4100 - accuracy: 0.8259 - val_loss: 0.3865 - val_accuracy: 0.8433
Epoch 5/100
7000/7000 [==============================] - 0s 54us/sample - loss: 0.3990 - accuracy: 0.8331 - val_loss: 0.3764 - val_accuracy: 0.8460
Epoch 6/100
7000/7000 [==============================] - 0s 58us/sample - loss: 0.3889 - accuracy: 0.8401 - val_loss: 0.3684 - val_accuracy: 0.8507
Epoch 7/100
7000/7000 [==============================] - 0s 55us

Epoch 56/100
7000/7000 [==============================] - 0s 50us/sample - loss: 0.3403 - accuracy: 0.8587 - val_loss: 0.3339 - val_accuracy: 0.8663
Epoch 57/100
7000/7000 [==============================] - 0s 58us/sample - loss: 0.3399 - accuracy: 0.8609 - val_loss: 0.3320 - val_accuracy: 0.8663
Epoch 58/100
7000/7000 [==============================] - 0s 50us/sample - loss: 0.3398 - accuracy: 0.8591 - val_loss: 0.3337 - val_accuracy: 0.8667
Epoch 59/100
7000/7000 [==============================] - 0s 51us/sample - loss: 0.3401 - accuracy: 0.8589 - val_loss: 0.3328 - val_accuracy: 0.8667
Epoch 60/100
7000/7000 [==============================] - 0s 53us/sample - loss: 0.3399 - accuracy: 0.8600 - val_loss: 0.3335 - val_accuracy: 0.8670
Epoch 61/100
7000/7000 [==============================] - 0s 48us/sample - loss: 0.3400 - accuracy: 0.8594 - val_loss: 0.3329 - val_accuracy: 0.8673
Epoch 62/100
7000/7000 [==============================] - 0s 51us/sample - loss: 0.3400 - accuracy: 0.8616

# Predict the results using 0.5 as a threshold

In [94]:
predications = model2.predict(X_test) #predicting

In [95]:
predications

array([[0.04467261],
       [0.02287269],
       [0.14361641],
       ...,
       [0.05559593],
       [0.01140234],
       [0.08059168]], dtype=float32)

In [96]:
y_pred = (predications > 0.5) #we are taking 0.5 as a threshold to predict whether they are likely to exit or not
print(y_pred)

[[False]
 [False]
 [False]
 ...
 [False]
 [False]
 [False]]


# Print the Accuracy score and confusion matrix

In [97]:
from sklearn.metrics import confusion_matrix

In [98]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[2332   84]
 [ 308  276]]


for calculating accuracy we are taking values which are correctly predicted, means values of the diagnols. We will sum tha values on the diagnol(2332+276)/3000 in this case.
3000 total test samples

In [101]:
print (((cm[0][0]+cm[1][1])*100)/y_test.shape, '% of testing data was classified correctly')

[86.93333333] % of testing data was classified correctly
